sdk/helper/compressutil/compress.go

```diff
@@ -141,10 +141,21 @@ func Compress(data []byte, config *CompressionConfig) ([]byte, error) {
 // If the first byte isn't a canary byte, then the utility returns a boolean
 // value indicating that the input was not compressed.
 func Decompress(data []byte) ([]byte, bool, error) {
+	bytes, _, notCompressed, err := DecompressWithCanary(data)
+	return bytes, notCompressed, err
+}
+
+// DecompressWithCanary checks if the first byte in the input matches the canary byte.
+// If the first byte is a canary byte, then the input past the canary byte
+// will be decompressed using the method specified in the given configuration. The type of compression used is also
+// returned. If the first byte isn't a canary byte, then the utility returns a boolean
+// value indicating that the input was not compressed.
+func DecompressWithCanary(data []byte) ([]byte, string, bool, error) {
```

Comment:
I'd prefer not to overload this function which already has a large number of return values, esp since currently callers only need one of these two things (the compression type, or the decompressed data).  How about instead adding a func like

```
func CompressionType(data []byte) string {
	if data == nil || len(data) < 2 {
		return ""
	}

	canary := data[0]

	switch canary {
	case CompressionCanaryGzip:
		return CompressionTypeGzip
	case CompressionCanaryLZW:
		return CompressionTypeLZW
	case CompressionCanarySnappy:
		return CompressionTypeSnappy
	case CompressionCanaryLZ4:
		return CompressionTypeLZ4
	}
	return ""
}
```
?

Reply:
I considered this approach before the last revision.

The problem is this approach would recreate a similar bug to #12746.  Any write to non-compressed content that happened to start with 'G', '4' like sys/token/salt would accidentally cause the write to be compressed.

The only way to avoid #12746 is during a write, verify the existing data is actually compressed data by attempting decompression.  If one is going to do that, then it makes sense to find out get the compression type at the same time.

Reply:
Ah, good point.